### PR TITLE
[IMP] Use helper function instead of hardcoded test passwords.

### PR DESCRIPTION
**Pull request by amh-mw**
# Description of the issue/feature this PR addresses:

As an Odoo developer, it is common for me to install some set of modules and then immediately run the tests to make sure that they work together reasonably well. Given that many tests create users, installing any sort of auth_password_policy-derived module can break just about everything, as the test passwords used are not sufficiently complex. On the flip side, the developers of those tests don't generally care about passwords; they just need users with specific group permissions.

# Current behavior before PR:

Tests use an assortment of hardcoded passwords that may not have sufficient complexity.

# Desired behavior after PR is merged:

New helper function `new_test_pass` deterministically returns a sufficiently complex password and developers no longer have to concern themselves with test passwords for most authentication scenarios. Existing test function `authenticate` no longer explicitly requires password.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Julien00859** [comment]
Hello world!

File: addons/analytic/tests/test_analytic_account.py [hunk 1]
```diff
@@ -31,7 +31,6 @@ def setUpClass(cls):
         user = cls.env['res.users'].create({
```
Comment (Julien00859) [inline addons/analytic/tests/test_analytic_account.py:31]: Couldn't `new_test_user` be used here too?

File: addons/web/tests/test_login.py [hunk 2]
```diff
@@ -18,11 +18,11 @@ def setUp(self):
         self.opener = Opener(self.env.cr)
         self.opener.cookies.set('session_id', self.session.sid, domain=HOST, path='/')
 
-    def login(self, username, password, csrf_token=None):
+    def login(self, username, password=None, csrf_token=None):
```
Comment (Julien00859) [inline addons/web/tests/test_login.py:21]: This test module is explicitly testing the login page and this method is an helper for the other tests, can you restore this method signature and use `new_test_pass` in the various test cases instead?

File: addons/website_livechat/tests/test_livechat_request.py [hunk 3]
```diff
@@ -27,7 +27,6 @@ def test_cancel_chat_request_on_visitor_demand(self):
             'name': 'Operator Marc',
             'login': 'operator_b',
             'email': 'operatormarc@example.com',
-            'password': "operatormarc",
             'livechat_username': "Marco'r El",
         })
```
Comment (Julien00859) [inline addons/website_livechat/tests/test_livechat_request.py:31]: Couldn't `new_test_user` be used here too?

File: odoo/tests/common.py [hunk 4]
```diff
@@ -1637,6 +1651,8 @@ def authenticate(self, user, password):
             # than this transaction.
             self.cr.flush()
             self.cr.clear()
+            if not password:
```
Comment (Julien00859) [inline odoo/tests/common.py:1654]: Nitpicking but I don't want an empty string to behave weird
```suggestion
            if password is None:
```

**Review by Julien00859 — COMMENTED** [review]
There may be other places where we could replace `env['res.user'].create(...)` with `new_new_user`. We need to change the enterprise repository too, @odoo/rd-runbot shall I just create a `new_test_pass` branch on the enterprise repo? Would it work?

Otherwise those changes seems reasonable to me.

I would had done it myself, I would has added an hardcoded complexe and long password inside of `odoo.tests.common` and imported it in all tests modules. Would it work too?

**Comment by Xavier-Do** [comment]
I would suggest:

```python
    def _check_password_policy(self, passwords):
        if odoo.modules.module.current_test:
            return
```

But this implied modifying code to make the test works. This is not the only place were we would be doing that and it works, but not sure @xmo-odoo will agree.

Another solution is to patch _check_password_policy in the TrasactionCase setUpClass. Only common.py is modified but globally and a patcher is added in any case, not ideal either.

A third solution is to hardcode a unique password for all users in common.py and use it everywhere. No need to make it dynamic per user. Can be substitute if needed. More diff but KISS

In any case I wouldn't add a method on res.users to generate the password. Overridable, yes, but adds an attribute on res.users for tests only, in that case I prefer the first solution.

**Comment by amh-mw** [comment]
> There may be other places where we could replace `env['res.user'].create(...)` with `new_new_user`.

I specifically didn't replace everything, simply trying to keep the pull request smaller, but I will go ahead and replace everything now.

**Comment by amh-mw** [comment]
How about something like this?

```python
def _check_password_policy(self, passwords):
    if odoo.modules.module.current_test and self.env.context.get('test_no_password_policy'):
        return
```

Then default the `test_no_password_policy` context to True in `new_test_user`? This should give the best of both worlds in that most test scenarios bypass password policy, but it is trivial to re-enable it.

**Comment by amh-mw** [comment]
> Another solution is to patch _check_password_policy in the TrasactionCase setUpClass.

Or I could patch new_test_pass downstream.

**Comment by Julien00859** [comment]
Don't work too much on this branch @amh-mw, we had a chat last week at work with xdo/xmo about your work and we weren't sure it was the ideal approach to the problem.

**Comment by amh-mw** [comment]
@Julien00859 Sorry to bother, but did you guys ever land on a direction you'd like to go with this problem? I have to keep patching tests to keep my own builds green and I'd just as soon expend the effort to build something lasting.

**Comment by Julien00859** [comment]
What was our decision again? Other than not deciding anything? @Xavier-Do @xmo-odoo

**Comment by tde-banana-odoo** [comment]
> What was our decision again? Other than not deciding anything? @Xavier-Do @xmo-odoo

We wait for the merge of https://github.com/odoo/odoo/pull/3989

**Comment by amh-mw** [comment]
> We wait for the merge of #3989

Is that the right issue? I fail to see how test users/passwords are related to a new XML-RPC endpoint?

**Comment by Julien00859** [comment]
@amh-mw tde is referring to a running gag we have at work, that other PR is the oldest PR of xmo that is still opened and we like to joke about that *one day* it'll be merged. I'm meeting in person with xdo to see if we can do something for you...

**Comment by Julien00859** [comment]
First of all, sorry, we forgot about this one... I'm now seeing those many conflicts and that's my fault because I'm the one who asked you to use "new_test_user" everywhere. Since last year we have been discussing about those passwords, there is a general feeling that we should just get rid of those passwords inside tests entirely. See #152378 and #152446 for prior work. We are challenging several ideas our side, trying to build a consensus among ourselves, this is slow and frankly we are nitpicking a lot. Our consensus about this PR, is that we rather take over and discuss the mater among ourselves, and not annoy the external contributors.

We too want to find a solution for those passwords in tests. We too want to make your experience better. It is just sad to say but we suck at contributing with the community and we rather implement a solution on our own.

Thank you for the time you put in this work. Also now than you know about the running gag, *hopefully this will be done before 3989*.

**Comment by Xavier-Do** [comment]
To complement juc answers, we usually try to avoid modifying to much code if it looks less necessary than the maintenance needed by the change. In this case the needs feels low for odoo devloppers since it works fine on our environment and is difficult to make work for all possible overrides of `_check_password_policy` outside of our code base.

To be more specific, in this case the proposed solution won't work if a policy asks for special char and uppercase and at least 6 not following number and ... Some website have password policy so strange that even strong password generator don't manage to pass them.

Even with the overridable password generator,  solution, we have no guarantee that it will work and it would require each module implementing `_check_password_policy` to add the companion method to ensure it works. But in this case, why not just disable the check `if config['test_enable']` (or similar). Less complexity, no modification needed in odoo.

This pr is also actually in conflict with the idea to remove password completely for tests users to improve performances. This will work fine with the current check since empty passwords are ignored, but if the override does not take care of this case, it won't.
What we can do is merge #152446, and hopefully all overrides of `_check_password_policy` should manage this case.

Another solution, that may be implemented, is a way to add patchers or setup for all test case from a module. This way, you could patch the method in your custom module to disable _check_password_policy when needed. This is actually already doable but there is no pretty way to do that.

**Comment by amh-mw** [comment]
@Julien00859 @Xavier-Do I'm not a huge fan of this pull request myself, but it unblocked me, so I pushed it upstream simply to highlight the need. That said, I would be totally fine with pinning my hopes on #152446, except that it appears to have been closed without merge?

**Comment by Julien00859** [comment]
@amh-mw You are right, it has been closed. They are thinking of reviving it, in one way or another.

**Pull request closed without merging**